docs/developing_iree/profiling.md

```diff
@@ -0,0 +1,217 @@
+# Profiling
+
+IREE [benchmarking](./benchmarking.md) gives us an overall view of the
+performance at a specific level of granularity. To understand performance
+details of the IREE system, one would need to
+[profile](https://en.wikipedia.org/wiki/Profiling_(computer_programming))
+IREE, in order to know how components at different levels function and interact.
+
+## Whole-system Profiling with Tracy
+
+IREE uses Tracy as the main tool to perform whole-system profiling.
+[Tracy](https://github.com/wolfpld/tracy) is a real-time, nanosecond resolution,
+remote telemetry, hybrid frame and sampling profiler. It can profile CPU, GPU,
+memory, locks, context switches, and many more.
+
+### Building Tracy
+
+To use tracing in IREE, you need to build IREE with following requirements:
+
+*   Turn `IREE_ENABLE_RUNTIME_TRACING` on.
+*   Add `-DNDEBUG` to `IREE_DEFAULT_COPTS`.
+*   Use Release/RelWithDebInfo build.
+
+For example:
+
+```shell
+export IREE_DEFAULT_COPTS='-DNDEBUG'
+cmake -B build/ \
+      -DIREE_ENABLE_RUNTIME_TRACING=ON \
+      -DCMAKE_BUILD_TYPE=RelWithDebInfo
+```
+
+The above compiles IREE with Tracy APIs so that IREE will stream profiling data
+back to Tracy when running. To be able to collect and analyze these data, you
+can either use GUI or CLI tools. Tracy profiler is the GUI tool. You can find
+the
+[Tracy manual](https://github.com/wolfpld/tracy/releases/download/v0.6.3/tracy.pdf)
+for more details on Tracy itself.
+
+To build the profiler on Linux, you will need to install some external
+libraries. Some Linux distributions will require you to add a `lib` prefix and a
+`-dev`, or `-devel` postfix to library names. For example, you might see the
+error:
+
+```
+Package glfw3 was not found in the pkg-config search path.
+```
+
+and then you could try to install `libglfw3-dev`.
+
+Instructions to build Tracy profiler:
+
+```shell
+cd third_party/tracy/profiler/build/unix
+make release
+```
+
+### Using Tracy
```

Comment:
A tip worth adding is that setting the TRACY_NO_EXIT=1 env var will keep the application you launch alive until tracy connects to it. This ensures that for quick apps (like run-module) that may exit in <500ms you have plenty of time to connect in the UI and pull the data over.

Reply:
Good point. Added.